pkg/ip/link_linux.go

```diff
@@ -220,28 +220,21 @@ func DelLinkByNameAddr(ifName string) ([]*net.IPNet, error) {
 	return out, nil
 }
 
-func SetHWAddrByIP(ifName string, ip4 net.IP, ip6 net.IP) error {
+// SetHardwareAddress will set the interface to a specified MAC address,
+// if the address is not assigned, we will generate a random CNI-OUI MAC
+// address for the interface.
+func SetHardwareAddress(ifName string, hwAddr net.HardwareAddr) error {
 	iface, err := netlink.LinkByName(ifName)
 	if err != nil {
 		return fmt.Errorf("failed to lookup %q: %v", ifName, err)
 	}
 
-	switch {
-	case ip4 == nil && ip6 == nil:
-		return fmt.Errorf("neither ip4 or ip6 specified")
+	if hwAddr == nil {
+		hwAddr = hwaddr.GenerateMAC()
+	}
 
-	case ip4 != nil:
-		{
-			hwAddr, err := hwaddr.GenerateHardwareAddr4(ip4, hwaddr.PrivateMACPrefix)
-			if err != nil {
-				return fmt.Errorf("failed to generate hardware addr: %v", err)
-			}
-			if err = netlink.LinkSetHardwareAddr(iface, hwAddr); err != nil {
-				return fmt.Errorf("failed to add hardware addr to %q: %v", ifName, err)
-			}
-		}
-	case ip6 != nil:
-		// TODO: IPv6
+	if err = netlink.LinkSetHardwareAddr(iface, hwAddr); err != nil {
```

Comment:
If the mac is randomly generated, can you retry if there's a duplicate? This is going to be sort of tough, since I don't think the kernel stops you from creating devices with duplicate MACs and you might need to scan every namespace.

Reply:
Unfortunately I don't think there is a good way to check MAC collision.. And all the plugins are generating MACs randomly now, so the collision risk is actually existing.
After the first three bytes are fixed, the probability of collision becomes larger. As my comments above, I did some tests on this and gave a **safe** usage suggestion.

Reply:
The kernel generates random addresses and does not check for collisions.

Reply:
Applying "Birthday Problem" analysis, 24 bits of randomness means that with 4096 containers we'd have a 50% chance of a collision.  This seems too low to be comfortable.

The current situation, letting Linux create random addresses, has >40 bits of randomness so far lower probability.

Reply:
mac collision only impacts if the containers are in the same layer 2 segments, right?
IIRC the max number of interfaces per linux bridge is 1024, another thing is stretched layer2 domains, but I don't think this applies here

Reply:
Emmm... This is public utility package, so maybe this will be called in other scenarios without 1024 max interfaces limit. But because of the limit, this is safe for CNI common plugins, right?

Reply:
yeah, I just wanted to point out that despite mac collision is a real problem, there are other more important limitations like this max number of interfaces per bridge or BUM traffic.
A layer 2 network with 4096 devices must be unmanageable, I actually don't know if all HW devices will be able to have a mac-table size that is able to host such amount of macs.

Reply:
yeap, got your point, the safe subnet prefix (22) is just for limiting cases, but in common cases, we don't need to care much about this.